Backport of:

From 9634b791c71c5eb9aae9b5b03326b14c454b8ac4 Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Thu, 15 Sep 2016 20:26:36 -0400
Subject: [PATCH] Memory allocate failure in AcquireQuantumPixels

This fix CVE-2016-8677

bug: https://github.com/ImageMagick/ImageMagick/issues/268
origin: https://github.com/ImageMagick/ImageMagick/commit/524349d2b3fed7fa0e53de2c908458474eb24418
bug-gentoo: https://blogs.gentoo.org/ago/2016/10/07/imagemagick-memory-allocate-failure-in-acquirequantumpixels-quantum-c/
bug-debian: https://bugs.debian.org/845206

(cherry picked from commit 524349d2b3fed7fa0e53de2c908458474eb24418)
---
 coders/tiff.c | 60 ++++++++++++++++++++++++++++++-----------------------------
 1 file changed, 31 insertions(+), 29 deletions(-)

Index: imagemagick-6.6.9.7/coders/tiff.c
===================================================================
--- imagemagick-6.6.9.7.orig/coders/tiff.c	2016-11-29 14:49:17.228512910 -0500
+++ imagemagick-6.6.9.7/coders/tiff.c	2016-11-29 14:49:17.224512865 -0500
@@ -1013,6 +1013,40 @@
       TIFFGetEXIFProperties(tiff,image);
     if (image_info->ping == MagickFalse)
       TIFFGetProfiles(tiff,image);
+    quantum_info=(QuantumInfo *) NULL;
+    if ((photometric == PHOTOMETRIC_PALETTE) &&
+        (pow(2.0,1.0*bits_per_sample) <= MaxColormapSize))
+      {
+        size_t
+          colors;
+
+        colors=(size_t) GetQuantumRange(bits_per_sample)+1;
+        if (AcquireImageColormap(image,colors) == MagickFalse)
+          {
+            TIFFClose(tiff);
+            ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+          }
+      }
+    if (units == RESUNIT_INCH)
+      image->units=PixelsPerInchResolution;
+    if (units == RESUNIT_CENTIMETER)
+      image->units=PixelsPerCentimeterResolution;
+    value=(unsigned short) image->scene;
+    (void) TIFFGetFieldDefaulted(tiff,TIFFTAG_PAGENUMBER,&value,&pages);
+    image->scene=value;
+    if (image_info->ping != MagickFalse)
+      {
+        if (image_info->number_scenes != 0)
+          if (image->scene >= (image_info->scene+image_info->number_scenes-1))
+            break;
+        goto next_tiff_frame;
+      }
+    status=SetImageExtent(image,image->columns,image->rows);
+    if (status == MagickFalse)
+      {
+        InheritException(exception,&image->exception);
+        return(DestroyImageList(image));
+      }
     /*
       Allocate memory for the image and pixel buffer.
     */
@@ -1124,39 +1158,6 @@
     if (image->matte != MagickFalse)
       (void) SetImageProperty(image,"tiff:alpha",
         associated_alpha != MagickFalse ? "associated" : "unassociated");
-    if ((photometric == PHOTOMETRIC_PALETTE) &&
-        (pow(2.0,1.0*bits_per_sample) <= MaxColormapSize))
-      {
-        size_t
-          colors;
-
-        colors=(size_t) GetQuantumRange(bits_per_sample)+1;
-        if (AcquireImageColormap(image,colors) == MagickFalse)
-          {
-            TIFFClose(tiff);
-            ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
-          }
-      }
-    if (units == RESUNIT_INCH)
-      image->units=PixelsPerInchResolution;
-    if (units == RESUNIT_CENTIMETER)
-      image->units=PixelsPerCentimeterResolution;
-    value=(unsigned short) image->scene;
-    (void) TIFFGetFieldDefaulted(tiff,TIFFTAG_PAGENUMBER,&value,&pages);
-    image->scene=value;
-    if (image_info->ping != MagickFalse)
-      {
-        if (image_info->number_scenes != 0)
-          if (image->scene >= (image_info->scene+image_info->number_scenes-1))
-            break;
-        goto next_tiff_frame;
-      }
-    status=SetImageExtent(image,image->columns,image->rows);
-    if (status == MagickFalse)
-      {
-        InheritException(exception,&image->exception);
-        return(DestroyImageList(image));
-      }
     method=ReadGenericMethod;
     if (TIFFGetField(tiff,TIFFTAG_ROWSPERSTRIP,&rows_per_strip) != 0)
       {
@@ -1690,7 +1691,8 @@
         if (status == MagickFalse)
           break;
       }
-    quantum_info=DestroyQuantumInfo(quantum_info);
+    if (quantum_info != (QuantumInfo *) NULL)
+      quantum_info=DestroyQuantumInfo(quantum_info);
   } while (status == MagickTrue);
   (void) TIFFSetWarningHandler(warning_handler);
   (void) TIFFSetErrorHandler(error_handler);
